File: dfs_bfs/단어변환.py
answer = int(1e9)
def solution(begin, target, words):
    global answer
    wdict = {}
    for word in words:
        wdict[word] = False
    if not target in words:
        return 0
    n = len(begin)
    
    def dfs(rst, tmp, target):
        global answer
        if tmp == target:
            answer = min(answer, rst)
            return
        wdict[tmp] = True
        for word in words:
            cnt = 0
            for i in range(n):
                if tmp[i] != word[i]:
                    cnt += 1
            if cnt > 1:
                continue
            elif cnt == 1:
                if not wdict[word]:
                    dfs(rst+1, word, target)
                    
    dfs(0, begin, target)
    return answer

File: dfs_bfs/단어변환.java
import java.util.Arrays;

class Solution {
    
    public static int answer;
    public static int words_size;
    public static int words_length;
    
    public int solution(String begin, String target, String[] words) {
        // 변환할 수 없는 경우
        if (!Arrays.asList(words).contains(target)) return 0;
        
        // 50개 이하의 단어가 있음
        answer = 51;
        
        words_size = words.length;
        words_length = words[0].length();
        
        boolean[] visited = new boolean[words_size];
        
        dfs(begin, target, words, visited, 0);
        return answer == 51 ? 0 : answer;
    }
    
    public void dfs(String begin, String target, String[] words, boolean[] visited, int count) {
        if (begin.equals(target)) {
            answer = Math.min(answer, count);
            return;
        }
        
        for (int i = 0; i < words_size; i++) {
            if (visited[i]) continue;
            
            int diff = 0;
            for (int j = 0; j < words_length; j++) {
                if (begin.charAt(j) != words[i].charAt(j)) {
                    diff++;
                }
                if (diff > 1) break;
            }
            
            if (diff == 1) {
                visited[i] = true;
                dfs(words[i], target, words, visited, count + 1);
                visited[i] = false;  
            }
            
        }
    }    
}
